Add analyzers for Microsoft.IdentityModel.Tokens (Wilson) Security Issues

TimHannMSFT opened the pull request:
Ensure key validation isn't explicitly disabled by setting the relevant setter on the validation object to false:
* RequireExpirationTime
* ValidateAudience
* ValidateIssuer
* ValidateLifetime

Ensure key validation isn't implicitly disabled by having the relevant delegate always return true:
* AudienceValidator
* LifetimeValidator

NOTE: There's another key implicit case for IssuerValidator, however the logic there is more complex and would require Data Flow Analysis (DFA). Since DFA has a different performance profile anyways, shipping these analyzers first which can be accomplished via static analysis. We also have CodeQL work in the pipe separately from this to address more complicated DFA flows.

<!--

Make sure you have read the contribution guidelines: 
- https://docs.microsoft.com/contribute/dotnet/dotnet-contribute-code-analysis#contribute-docs-for-caxxxx-rules
- https://github.com/dotnet/roslyn-analyzers/blob/main/GuidelinesForNewRules.md

If your Pull Request is doing one of the following:

- Adding a new diagnostic analyzer or a code fix
- Adding or updating resource strings used by analyzers and code fixes
- Updating analyzer package versions in [Versions.props](../eng/Versions.props)

Then, make sure to run `msbuild /t:pack /v:m` in the repository root; otherwise, the CI build will fail.

Note: Consider merging the PR base branch (`2.9.x`, `main`, or `release/*`) into your branch before you run the pack command to reduce merge conflicts.

-->

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=5420) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=5420) before we can accept your contribution.<br/><br/>:x: TimHannMSFT [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=5420)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn-analyzers?pullRequest=5420) it.</sub>

TimHannMSFT commented:
Addressing https://github.com/dotnet/roslyn-analyzers/issues/5419

TimHannMSFT commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,234 @@
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotAlwaysSkipTokenValidationInDelegates : DiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "CA5405"; // TODO: Somewhere in documentation I saw I need to create a CA5405.md file, don't recall where.
```
Found the doc I saw earlier: https://github.com/dotnet/roslyn-analyzers/blob/main/GuidelinesForNewRules.md

Will follow these rules for adding documentation for the two new rules.

dotpaul replied:
I believe it's here: [https://github.com/dotnet/docs/tree/main/docs/fundamentals/code-analysis/quality-rules](https://github.com/dotnet/docs/tree/main/docs/fundamentals/code-analysis/quality-rules)

TimHannMSFT commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/xlf/MicrosoftNetCoreAnalyzersResources.cs.xlf:
```diff
@@ -417,6 +417,21 @@
         <target state="translated">Toto přetížení metody Add je potenciálně nebezpečné, protože může překládat nebezpečné externí odkazy.</target>
         <note />
       </trans-unit>
+      <trans-unit id="DoNotAlwaysSkipTokenValidationInDelegates">
```
Not familiar with the localization process here. Didn't see any reference to this in documentation, let me know if there are other steps I need to do regarding this. Thanks.

dotpaul replied:
Localizers will localize later.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1266,4 +1266,22 @@
     <value>Ensure autogenerated class containing DataSet.ReadXml() is not used with untrusted data</value>
   </data>
+  <data name="DoNotDisableTokenValidationChecks" xml:space="preserve">
+    <value>Do Not Disable Token Validation Checks</value>
```
I think this should be sentence casing, ie `Do not disable token validation checks`

TimHannMSFT replied:
Thanks! I was patterning off of "DoNotDisableCertificateValidation::Do Not Disable Certificate Validation" but looking again that seems like an anomaly. Will update this and the other.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1266,4 +1266,22 @@
+    <value>TokenValidationParameters.{0} should not be set to false as it disables important validation.</value>
+  </data>
+  <data name="DoNotAlwaysSkipTokenValidationInDelegates" xml:space="preserve">
+    <value>Do Not Always Skip Token Validation In Delegates</value>
```
Same casing comment.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,234 @@
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
```
Can you rename this to `context`? (Intentionally reusing and shadowing the existing name)

TimHannMSFT replied:
What benefit does this offer (and below)? I'm following a pattern from another analyzer here and it seems to help legibility to make it obvious that these variables belong to different scopes.

Youssef1313 replied:
@TimHannMSFT One of the benefits is that this prevents accidentally using a cancellation token from an outer context. I don't know if there are other benefits. But @sharwell will know.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,234 @@
+                    var tokenValidationParameters = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftIdentityModelTokensTokenValidationParameters);
+
+                    if ((microsoftIdentityModelTokensAudienceValidatorTypeSymbol == null
+                        || ienumString == null
```
`GetSpecialType` should never return null.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,234 @@
+                    }
+
+                    compilationStartAnalysisContext.RegisterOperationAction(
+                        (OperationAnalysisContext operationAnalysisContext) =>
```
Same about renaming.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,234 @@
+                return false;
+            }
+
+            if (!parameters[0].Type.Equals(iEnumerableString)
+                || !parameters[1].Type.Equals(securityToken)
+                || !parameters[2].Type.Equals(tokenValidationParameters))
```
```suggestion
            if (!parameters[0].Type.Equals(iEnumerableString, SymbolEqualityComparer.Default)
                || !parameters[1].Type.Equals(securityToken, SymbolEqualityComparer.Default)
                || !parameters[2].Type.Equals(tokenValidationParameters, SymbolEqualityComparer.Default))
```

TimHannMSFT replied:
No such overload in 2.9.x

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotDisableTokenValidationChecks.cs:
```diff
@@ -0,0 +1,91 @@
+                    {
+                        var property = propertyReferenceOperation.Property;
+
+                        if (property.ContainingType.Equals(tokenValidationParamsTypeSymbol) &&
```
```suggestion
                        if (property.ContainingType.Equals(tokenValidationParamsTypeSymbol, SymbolEqualityComparer.Default) &&
```

TimHannMSFT replied:
Looks like this overload doesn't exist in 2.9.x

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Security/DoNotDisableTokenValidationChecksTests.cs:
```diff
@@ -0,0 +1,230 @@
+            await csharpTest.RunAsync();
+        }
+
+        private static DiagnosticResult GetCSharpResultAt(int line, int column)
+            => VerifyCS.Diagnostic()
+                .WithLocation(line, column);
```
@Evangelink Didn't you ban this overload before? 😨 

@TimHannMSFT Please remove this helper method and use the markup syntax instead. This works by surrounding the code where the diagnostic is expected between `[|` and `|]`

Example:

https://github.com/dotnet/roslyn-analyzers/blob/762b08948cdcc1d94352fba681296be7bf474dd7/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs#L478

TimHannMSFT replied:
I can certainly have a look, seems this is used in quite a few of the other test classes though:

DoNotDisableCertificateValidationTests.cs, DoNotDisableHTTPHeaderCheckingTests.cs, etc

Youssef1313 replied:
@TimHannMSFT Yes it's still used in some tests. But I believe it's a goal to get rid of them as time goes on. That's why such overloads were banned in https://github.com/dotnet/roslyn-analyzers/pull/4631. I'm not sure why there is no errors here for using the banned symbols.

dotpaul replied:
`WithLocation()` isn't banned in the `2.9.x` branch (guess there wasn't an altnernative?). I'd prefer sticking with the old style in both `2.9.x` and `main` to keep merges simpler when there are fixes. We can add `#pragma warning disable RS0030` in `main` like we do for other security analyzer tests.

TimHannMSFT replied:
Sounds good, I went through and did this before seeing Paul's comment but I stashed those changes and will leave as is for now.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegatesTests.cs:
```diff
@@ -0,0 +1,602 @@
+                {
+                    Sources = { source2 },
+                },
+                SolutionTransforms =
+                {
+                    (solution, projectId) =>
+                    {
+                        var sideProject = solution.AddProject("DependencyProject", "DependencyProject", LanguageNames.CSharp)
+                            .AddDocument("Dependency.cs", source1).Project
+                            .AddMetadataReferences(solution.GetProject(projectId).MetadataReferences)
+                            .WithCompilationOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+                        return sideProject.Solution.GetProject(projectId)
+                            .AddProjectReference(new ProjectReference(sideProject.Id))
+                            .Solution;
+                    }
+                }
```
Would something like this work?

https://github.com/dotnet/roslyn-analyzers/blob/762b08948cdcc1d94352fba681296be7bf474dd7/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Security/DoNotDisableCertificateValidationTests.cs#L197-L211

TimHannMSFT replied:
It appears this syntax isn't yet supported in the 2.9.x branch @dotpaul recommended I target. Will leave as is which matches the implementation for the class you referenced in 2.9.x branch.

dotpaul commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,234 @@
+                                    var targetOperations = blockOperation.Descendants().ToImmutableArray().WithoutFullyImplicitOperations();
+                                    alwaysReturnTrue = AlwaysReturnTrue(targetOperations);
+                                    break;
+                            }
```
I like to add a default case, and just 

```csharp
    default:
        Debug.Fail("Unhandled OperationKind " + delegateCreationOperation.Target.Kind);
        break;
```

dotpaul commented:
Other than the casing in localizable strings that @Youssef1313 mentioned, LGTM

TimHannMSFT commented:
I believe I've addressed all the PR feedback and I've pushed a PR to the docs branch to address those requirements as well (https://github.com/dotnet/docs/pull/25911)

henrik-me commented:
Were these rules run on our samples already? If not I suggest running those on our web app and web api samples.

TimHannMSFT commented on src/Test.Utilities/AdditionalMetadataReferences.cs:
```diff
@@ -37,6 +37,9 @@ public static class AdditionalMetadataReferences
             .AddPackages(ImmutableArray.Create(new PackageIdentity("Newtonsoft.Json", "10.0.1")));
 
+        public static ReferenceAssemblies DefaultWithWilson { get; } = Default
+            .AddPackages(ImmutableArray.Create(new PackageIdentity("Microsoft.IdentityModel.Tokens", "6.12.0")));
```
My understanding is that this is just what binaries are referenced in the test samples. It's all based off of the namespaces and type names which should be static across versions.

TimHannMSFT commented:
> Were these rules run on our samples already? If not I suggest running those on our web app and web api samples.

The unit test files run the rules against a number of samples. Definitely open to other ideas/permutations on those. See DoNotDisableTokenValidationChecksTests.cs and DoNotAlwaysSkipTokenValidationInDelegatesTests.cs for the sample code the unit tests run against.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1266,4 +1266,22 @@
   <data name="DataSetReadXmlAutogeneratedTitle" xml:space="preserve">
     <value>Ensure autogenerated class containing DataSet.ReadXml() is not used with untrusted data</value>
   </data>
+  <data name="DoNotDisableTokenValidationChecks" xml:space="preserve">
```
We usually go with `Title` for this case.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,240 @@
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext context) =>
```
[minor][cosmetic] Parenthesis aren't necessary.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,240 @@
+                    }
+
+                    context.RegisterOperationAction(
+                        (OperationAnalysisContext context) =>
```
[minor][cosmetic] Parenthesis aren't necessary.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,240 @@
+                        (OperationAnalysisContext context) =>
+                        {
+                            var delegateCreationOperation =
+                                (IDelegateCreationOperation)context.Operation;
```
[minor][cosmetic] Line split isn't necessary.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotAlwaysSkipTokenValidationInDelegates.cs:
```diff
@@ -0,0 +1,240 @@
+
+            foreach (var descendant in operations)
+            {
+                if (descendant.Kind == OperationKind.Return)
```
[minor][cosmetic] You could revert this condition to reduce nesting.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotDisableTokenValidationChecks.cs:
```diff
@@ -0,0 +1,91 @@
+    public sealed class DoNotDisableTokenValidationChecks : DiagnosticAnalyzer
+    {
+        // Set of properties on Microsoft.IdentityModel.Tokens.TokenValidationParameters which shouldn't be set to false.
+        private readonly string[] PropertiesWhichShouldNotBeFalse = new string[]
```
You may want to consider using `ImmutableArray`, perf is the same as array but you have the immutability benefice.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotDisableTokenValidationChecks.cs:
```diff
@@ -0,0 +1,91 @@
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var tokenValidationParamsTypeSymbol = compilation.GetOrCreateTypeByMetadataName(
```
[minor] You could use the `TryGetOrCreate...` version rather than checking for `null` after.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotDisableTokenValidationChecks.cs:
```diff
@@ -0,0 +1,91 @@
+                {
+                    var simpleAssignmentOperation = (ISimpleAssignmentOperation)context.Operation;
+
+                    if (simpleAssignmentOperation.Target is IPropertyReferenceOperation propertyReferenceOperation)
```
[minor] You could revert this condition to reduce nesting.

dotpaul commented:
> > Were these rules run on our samples already? If not I suggest running those on our web app and web api samples.
> 
> The unit test files run the rules against a number of samples. Definitely open to other ideas/permutations on those. See DoNotDisableTokenValidationChecksTests.cs and DoNotAlwaysSkipTokenValidationInDelegatesTests.cs for the sample code the unit tests run against.

Not blocking this PR, but if you have a chance, it'd be good to run the analyzers against a large real world code base and look for AD0001 errors (unhandled exceptions) or false positives you'd want to address.

mafurman commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1266,4 +1266,22 @@
     <value>Ensure autogenerated class containing DataSet.ReadXml() is not used with untrusted data</value>
   </data>
+  <data name="DoNotDisableTokenValidationChecks" xml:space="preserve">
+    <value>Do not disable token validation checks</value>
```
nit: put period at the end of the sentence for consistency

dotpaul replied:
Don't need punctuation for titles (and probably should rename the string to end with "Title" like @Evangelink suggested)

mafurman commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1266,4 +1266,22 @@
+    <value>Do not always skip token validation in delegates</value>
+  </data>
+  <data name="DoNotAlwaysSkipTokenValidationInDelegatesDescription" xml:space="preserve">
+    <value>By setting critical TokenValidationParameter validations delegates to true, important authentication safeguards are disabled which can lead to tokens from any issuer or expired tokens being wrongly validated.</value>
```
nit: validations -> validation

mafurman commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1266,4 +1266,22 @@
+    <value>By setting critical TokenValidationParameter validations delegates to true, important authentication safeguards are disabled which can lead to tokens from any issuer or expired tokens being wrongly validated.</value>
+  </data>
+  <data name="DoNotAlwaysSkipTokenValidationInDelegatesMessage" xml:space="preserve">
+    <value>The {0} is set to a function that is always returning true. By setting the validation delegate you are overriding default validation and by always returning true, this validation is completely disabled.</value>
```
nit: should have a comma after "By setting the validation delegate"

mafurman commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/xlf/MicrosoftNetCoreAnalyzersResources.cs.xlf:
```diff
@@ -417,6 +417,21 @@
+      </trans-unit>
+      <trans-unit id="DoNotAlwaysSkipTokenValidationInDelegatesMessage">
+        <source>The {0} is set to a function that is always returning true. By setting the validation delegate you are overriding default validation and by always returning true, this validation is completely disabled.</source>
+        <target state="new">The {0} is set to a function that is always returning true. By setting the validation delegate you are overriding default validation and by always returning true, this validation is completely disabled.</target>
```
similar nits to those above are applicable in this file as well

dotpaul reviewed (approved):
Thank you, @TimHannMSFT !

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/xlf/MicrosoftNetCoreAnalyzersResources.cs.xlf:
```diff
@@ -552,6 +567,21 @@
+        <note />
+      </trans-unit>
+      <trans-unit id="DoNotDisableTokenValidationChecksMessage">
+        <source>TokenValidationParameters.{0} should not be set to false as it disables important validation.</source>
```
@Evangelink This shouldn't end in a period right?

TimHannMSFT replied:
All the message elements I saw ended with a period. The titles had none and I tried to match that style there too. If there's a different consensus on style I can always push an update.

Youssef1313 replied:
@TimHannMSFT, I believe the message should only end in a period when multi-statement, and doesn't if single-statement.

There is already an analyzer for that, but it either got confused by the `.{0}` and considered it a multi-statement or it's not working for some other reason.

The rule for that is:

https://github.com/dotnet/roslyn-analyzers/blob/aac24788f059850f12978bc50b80444607e1f21c/src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx#L505-L507

Youssef1313 replied:
The analyzer is outdated in 2.9.x branch.
